### ISSUE TITLE
`unnecessary_to_owned` doesn't take type changes into account

### ISSUE THREAD

**Issue by mikerite** [rust-lang/rust-clippy#9351]
### Summary

`unnecessary_to_owned` suggests changing the type of an input but doesn't check if this type is used in other places. This can lead to suggestions that won't compile due to mismatched type errors.

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
#![allow(unused_variables)]
use std::ops::Deref;
use std::path::PathBuf;

fn main() {
    let path = std::path::Path::new("x");
    let x: PathBuf = require_deref_path(path.to_owned());
    // let x: PathBuf = require_deref_path(path);
}

fn require_deref_path<T: Deref<Target = std::path::Path>>(x: T) -> T {
    x
}
```

I saw this happen:

```
warning: unnecessary use of `to_owned`
 --> src/main.rs:7:41
  |
7 |     let x: PathBuf = require_deref_path(path.to_owned());
  |                                         ^^^^^^^^^^^^^^^ help: use: `path`
  |
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
```

Applying the suggestion results in this error
```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
  --> src/main.rs:8:41
   |
8  |     let x: PathBuf = require_deref_path(path);
   |                      ------------------ ^^^^- help: try using a conversion method: `.to_path_buf()`
   |                      |                  |
   |                      |                  expected struct `PathBuf`, found `&Path`
   |                      arguments to this function are incorrect
   |
note: function defined here
  --> src/main.rs:11:4
   |
11 | fn require_deref_path<T: Deref<Target = std::path::Path>>(x: T) -> T {
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d8f0c9b97e862a49063b2a08d6113205)

I expected to see this happen:
No warning.

### Version

```text
Clippy 0.1.65 (2022-08-16 86c6ebe)
```

### Additional Labels

@rustbot +label I-suggestion-causes-error
